src/base/string.h

```diff
@@ -0,0 +1,132 @@
+/*
+ * Copyright (C) 2021 by the Widelands Development Team
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ */
+
+#ifndef WL_BASE_STRING_H
+#define WL_BASE_STRING_H
+
+#include <algorithm>
+#include <functional>
+#include <set>
+#include <string>
+#include <vector>
+
+/** Split a string into substrings at all occurrences of any of the given delimiters. */
+void split(std::vector<std::string>& result, const std::string&, const std::set<char>&);
+
+/** Convert a string to lowercase. */
+inline std::string to_lower(const std::string& str) {
+	std::string result = str;
+	std::transform(result.begin(), result.end(), result.begin(), tolower);
+	return result;
+}
+
+/** Check whether two strings are equal regardless of case. */
+inline bool iequals(const std::string& a, const std::string& b) {
+	return to_lower(a) == to_lower(b);
+}
+
+/**
+ * Check whether `str` starts with / ends with / contains `test`.
+ * Can be case-sensitive or -insensitive depending on the last argument, default is sensitive.
+ */
+inline bool
+starts_with(const std::string& str, const std::string& test, bool case_sensitive = true) {
```

Comment:
For case-sensitive comparison, you should not call a separate function but just use the built-in `==` operator:
```
iequals(str1, str2)  // Insensitive
str1 == str2         // Sensitive
```

Boost has an `i`-prefixed version for each of its functions (e.g. `boost::ends_with` and `boost::iends_with`), but a unified function with a parameter is more readable, IMHO, and it's also more common in the sense that the STL defines only a small number of methods with sometimes dozens of overloads for multiple purposes. Complexity-wise, the ternary operator can be optimised out in release builds by the compiler (as it's inlined and always a boolean literal) so there is no difference in performance either.

Reply:
Actually, I just realized that by pulling the case-insensitive variant out into separate functions I can greatly reduce their complexity. Will do so…

Reply:
…but this would make the code so hard to read (102 lines of low-level C-string magic vs 20 lines of high-level C++) that I'd say let's choose readability over efficiency in this case; especially as these functions are not used a lot, and never in performance-critical or even much-used code paths.